src/fsharp/service/ServiceAnalysis.fs

```diff
@@ -70,90 +85,117 @@ module UnusedOpens =
+                openedGroup.OpenedModules |> List.exists (fun openedEntity ->
+
+                    symbolUses1 |> Array.exists (fun symbolUse ->
+                        rangeContainsRange openStatement.AppliedScope symbolUse.RangeAlternate &&
+                        match symbolUse.Symbol with
+                        | :? FSharpMemberOrFunctionOrValue as f ->
+                            match f.DeclaringEntity with
+                            | Some ent when ent.IsNamespace || ent.IsFSharpModule -> ent.IsEffectivelySameAs openedEntity.Entity
+                            | _ -> false
+                        | _ -> false) || 
+
+                    symbolUses2 |> Array.exists (fun symbolUse ->
+                        rangeContainsRange openStatement.AppliedScope symbolUse.RangeAlternate &&
+                        openedEntity.RevealedSymbolsContains symbolUse.Symbol)))
+
+        // Return them as interim used entities
+        newlyUsedOpenedGroups |> List.collect (fun openedGroup -> 
+                openedGroup.OpenedModules |> List.map (fun x -> { Module = x.Entity; AppliedScope = openStatement.AppliedScope }))
                                           
-                match openStatements with
-                | os :: xs ->
-                    match getUsedModules os with
-                    | [] -> filterInner (os :: acc) xs usedModules
-                    | um -> filterInner acc xs (um @ usedModules)
-                | [] -> List.rev acc
-            
-            filterInner [] openStatements []
+    /// Incrementally filter out the open statements one by one. Filter those whose contents are referred to somewhere in the symbol uses.
+    /// Async to allow cancellation.
+    let rec filterOpenStatementsIncremental symbolUses (openStatements: OpenStatement list) (usedModules: UsedModule list) acc = 
+        async { 
+            match openStatements with
+            | openStatement :: rest ->
+                match getUsedModules symbolUses usedModules openStatement with
+                | [] -> 
+                    // The open statement has not been used, include it in the results
+                    return! filterOpenStatementsIncremental symbolUses rest usedModules (openStatement :: acc)
+                | moreUsedModules -> 
+                    // The open statement has been used, add the modules which are already known to be used to the list of things we don't need to re-check
+                    return! filterOpenStatementsIncremental symbolUses rest (moreUsedModules @ usedModules) acc
+            | [] -> return List.rev acc
+        }
+
+    /// Filter out the open statements whose contents are referred to somewhere in the symbol uses.
+    /// Async to allow cancellation.
+    let filterOpenStatements symbolUses openStatements =
+        async {
+            let! results = filterOpenStatementsIncremental symbolUses openStatements [] []
+            return results |> List.map (fun os -> os.Range)
```

Comment:
This comment isn't about the PR itself but is about ranges this feature deals with.
I'd in favour of having `SynMemberDefn.Open` tree nodes here (and other similar cases where a range is returned to a client) instead of lid ranges if these nodes had ranges of the whole expressions instead of its long identifiers. Furthermore, it'd be great if other declarations (namespace, type, etc) ranges included starting keywords (or even pointed to a token `pos`).

Reply:
That makes sense (but yes, is orthogonal)

Reply:
This https://github.com/Microsoft/visualfsharp/pull/4106 is related. We should have discussions about that in there.

Also, `SynMemberDefn.Open` doesn't seem to be used anywhere that I can find in VisualFSharp.

Reply:
@TIHan we use it in Rider along with other types from AST.

Reply:
Correction, I don't see SymMemberDefn.Open being constructed in VisualFsharp.

Reply:
@TIHan it looks I copied the wrong type name in the first message. It should have been `SynModuleDecl.Open`, thank you.